sktime/transformations/compose.py

```diff
@@ -0,0 +1,502 @@
+# -*- coding: utf-8 -*-
+"""Meta-transformers for building composite transformers."""
+# copyright: sktime developers, BSD-3-Clause License (see LICENSE file)
+
+from sklearn.base import clone
+
+from sktime.base import _HeterogenousMetaEstimator
+from sktime.transformations.base import BaseTransformer
+
+__author__ = ["fkiraly"]
+__all__ = ["TransformerPipeline"]
+
+
+class TransformerPipeline(BaseTransformer, _HeterogenousMetaEstimator):
+    """Pipeline of transformers compositor.
+
+    The `TransformerPipeline` compositor allows to chain transformers.
+    The pipeline is constructed with a list of sktime transformers,
+        i.e., estimators following the BaseTransformer interface.
+    The list can be unnamed - a simple list of transformers -
+        or string named - a list of pairs of string, estimator.
+
+    For a list of transformers `trafo1`, `trafo2`, ..., `trafoN`,
+        the pipeline behaves as follows:
+    `fit` - changes state by running `trafo1.fit_transform`, `trafo2.fit_transform`, etc
+        sequentially, with `trafo[i]` receiving the output of `trafo[i-1]`
+    `transform` - result is of executing `trafo1.transform`, `trafo2.transform`, etc
+        with `trafo[i].transform` input = output of `trafo[i-1].transform`,
+        and returning the output of `trafoN.transform`
+    `inverse_transform` - result is of executing `trafo[i].inverse_transform`,
+        with `trafo[i].inverse_transform` input = output `trafo[i-1].inverse_transform`,
+        and returning the output of `trafoN.inverse_transform`
+    `update` - changes state by chaining `trafo1.update`, `trafo1.transform`,
+        `trafo2.update`, `trafo2.transform`, ..., `trafoN.update`,
+        where `trafo[i].update` and `trafo[i].transform` receive as input
+            the output of `trafo[i-1].transform`
+
+    `get_params`, `set_params` uses `sklearn` compatible nesting interface
+        if list is unnamed, names are generated as names of classes
+        if names are non-unique, `f"_{str(i)}"` is appended to each name string
+            where `i` is the total count of occurrence of a non-unique string
+            inside the list of names leading up to it (inclusive)
+
+    `TransformerPipeline` can also be created by using the magic multiplication
+        on any transformer, i.e., any estimator inheriting from `BaseTransformer`
+            for instance, `my_trafo1 * my_trafo2 * my_trafo3`
+            will result in the same object as  obtained from the constructor
+            `TransformerPipeline([my_trafo1, my_trafo2, my_trafo3])`
+        magic multiplication can also be used with (str, transformer) pairs,
+            as long as one element in the chain is a transformer
+
+    Parameters
+    ----------
+    transformers : list of sktime transformers, or
```

Comment:
I would prefer to only accept tuple input as in our other pipeline classes and as in `sklearn`. I think now it makes the interface more confusing and its not. Also in the zen of Python, it says:

> There should be one-- and preferably only one --obvious way to do it.

😃

Reply:
1 Beautiful is better than ugly.
2 Simple is better than complex.
3 Flat is better than nested.
4 Readability counts.
-> compare `my_trafo * my_trafo2` with `TransformerPipeline(steps=[("mytrafo", mytrafo), ("mytrafo2", mytrafo2)])`